rfc/0015-multi-file-jayvee/0015-multi-file-jayvee.md

```diff
@@ -0,0 +1,116 @@
+  - DISCUSSION: The RFC is open for discussion. Usually, we open a PR to trigger discussions.
+  - ACCEPTED: The RFC was accepted. Create issues to prepare implementation of the RFC.
+  - REJECTED: The RFC was rejected. If another revision emerges, switch to status DRAFT.
+-->
+
+## Summary
+
+This RFC introduces the possibility of distributing a Jayvee program over multiple files.
+This feature will foster reuse of valuetypes, blocks, and other elements.
+Inherent to this feature is a concept of how scoping and naming is handled for nested structures.
+
+
+## Motivation
+
+Right now, Jayvee users can only pack their whole Jayvee model into one file. 
+The challenge is two-fold:
+1. Larger projects will become unmaintainable quite quickly, as the elements cannot be organized into multiple files.
+2. Without distribution to multiple files, there is no possibility to reuse models of other projects.
+
+For example, we might be enable to build libraries of valuetypes that can be reused across multiple projects instead of copying the code.
+
+## Explanation
+
+### Exporting elements
+
+For exporting elements, I propose introducing a new concept called `libraries` instead of explicitly modeling an import or export per element.
+A `library` can inhibit `Valuetype`s, `Block`s, `BlockType`s,  `Constraint`s, and `Transform`s.
+
+**Example library**
+```
+library MyDomainLibrary {
+  valuetype MyDomainSpecificValuetype {
+    // ... details of valuetype
+  }
+
+  // ... possibly more elements
+}
+```
+
+Libraries are always "exported" and can be "imported" into other files.
+
+
+### Visibility of elements in a file
+
+By introducing the concept of `libraries`, most elements can be defined on three levels in a Jayvee file:
+1. On the root level of the file
+2. Within a pipeline
+3. (new) Within a library
+
+The name of an element is given by its definition. The **qualified name** is constructed by prepending container structures in this pattern: `<container name>.<element name>`, e.g., `MyDomainLibraryMyDomainLibrary.MyDomainSpecificValuetype`.
+
+**Access paths:**
+- root level elements can access 
+  - root level elements by their name
+  - and elements of libraries by their qualified name
+- elements within a library can access
+  - root level elements by their name
+  - elements within the same library by their name
+  - and elements of other libraries by their qualified name
+- elements within a pipeline can access
+  - root level elements by their name
+  - elements within the same pipeline by their name
+  - and elements of libraries by their qualified name
+
+**No-access paths:**
+- elements within a pipeline cannot be referenced by outside elements
+
+
+### Importing elements
+
+Only `libraries` and their elements can be imported into other files. Elements on the root level of a file or within a pipeline cannot be imported.
+
+```
+import './path/to/jv-file.jv'; // imports all libraries of the file
```

Comment:
I'd immediately talk of a `location` here by the way. We can by default map it to files but I feel it would be good to medium term have this more open to e.g. loading libraries from the web.